library/psa_crypto.c

```diff
@@ -2519,41 +2521,41 @@ psa_status_t psa_hash_abort( psa_hash_operation_t *operation )
              * in use. It's ok to call abort on such an object, and there's
              * nothing to do. */
             break;
-#if defined(MBEDTLS_MD2_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_MD2)
         case PSA_ALG_MD2:
             mbedtls_md2_free( &operation->ctx.md2 );
             break;
 #endif
-#if defined(MBEDTLS_MD4_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_MD4)
         case PSA_ALG_MD4:
             mbedtls_md4_free( &operation->ctx.md4 );
             break;
 #endif
-#if defined(MBEDTLS_MD5_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_MD5)
         case PSA_ALG_MD5:
             mbedtls_md5_free( &operation->ctx.md5 );
             break;
 #endif
-#if defined(MBEDTLS_RIPEMD160_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_RIPEMD160)
         case PSA_ALG_RIPEMD160:
             mbedtls_ripemd160_free( &operation->ctx.ripemd160 );
             break;
 #endif
-#if defined(MBEDTLS_SHA1_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_1)
         case PSA_ALG_SHA_1:
             mbedtls_sha1_free( &operation->ctx.sha1 );
             break;
 #endif
-#if defined(MBEDTLS_SHA256_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_256) || \
```

Comment:
With this solution, if MBEDTLS_PSA_BUILTIN_ALG_SHA_224 is defined but not MBEDTLS_PSA_BUILTIN_ALG_SHA_256, the function will not return in error when called for algorithm MBEDTLS_PSA_BUILTIN_ALG_SHA_256. This is probably not a big deal but strictly speaking this is not correct. Furthermore this is not aligned with the way it is done in `psa_hash_setup`. Thus I would prefer to align with what is done in `psa_hash_setup`:
```c
#if defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_224)
         case PSA_ALG_SHA_224:
            mbedtls_sha256_free( &operation->ctx.sha256 );
            break;
#endif
#if defined(MBEDTLS_PSA_BUILTIN_ALG_SHA_256)
         case PSA_ALG_SHA_256:
            mbedtls_sha256_free( &operation->ctx.sha256 );
            break;
#endif
```

Reply:
Unsupported cases are unreachable: `psa_hash_setup` sets the `operation->alg` and it won't set it to an unsupported algorithm. The only advantage of being strict here is saving a few bytes of code when only one of the two algorithms is enabled (because there's one less target value for the switch). But never mind.